### ISSUE TITLE
Wrong calculation of running Template agents when using the same image on multiple templates

### ISSUE THREAD

**Issue by ericcitaire** [jenkinsci/docker-plugin#816]
- [X] docker-plugin version : **1.2.1**
 - [X] jenkins version : **2.249.2**
 - [X] docker engine version : **19.03.13**

In [`com.nirima.jenkins.plugins.docker.DockerCloud` in `canAddProvisionedSlave()`](https://github.com/jenkinsci/docker-plugin/blob/c3a38d7034132cb407832a4ae0372f44339a2b03/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java#L650), an estimation of running template agents is done.

This is done by counting the running containers using the same image as the template image, which is wrong since multiple templates of the same cloud can use the same Docker image. In my case, I have 2 templates with the same Docker image, but different container capacities and different agent labels (for limiting the builds).

It may be fixed by adding a template label to each container and add a filter on this label to count running template agents?

**Comment by pjdarton** [comment]
Yup, you're right - using the image name (only) is insufficiently-unique to count templates.
FYI it also gets confused when jobs specify their own job-specific template.
...and this false assumption is also made when the code tries to count the number of containers "under construction" too (which it tries to keep track of in order to reduce the amount of over-provisioning, as Jenkins keeps asking for more until they've come online, at which point we might have lots "under contruction" and get an excess).
i.e. This bug has been around for a long long time and nobody's (yet) plucked up the courage to fix it.

It might be easier these days than it used to be - now we have template names, which could be used to help differentiate ... except we don't insist folks set them so they can end up all looking the same that way too.

TL;DR: Yup, it's a bug.  A PR to fix it would be welcomed.
